### PR TITLE
Custom fonts

### LINKED ISSUE #1646: Font Customization
**Just a suggestion. Would it be possible for you to add an addon that allows you to customize the font on the website?**

**It's nothing that is really useful, but I think it would make it look better to the eyes.**

### PR DESCRIPTION
**Pull request by ErrorGamer2000**
Resolves #1646

### Changes

Added the `Customizable fonts` addon. (The second version) This addon has options for the Navbar, Main, Header, Editor, and Block

### Reason for changes

It seems like a good addon . . .

### Tests

https://user-images.githubusercontent.com/70162741/118883103-1e7b0400-b8b2-11eb-8988-68d7e43ea5f6.mp4

Just imagine the popup there please, my recorder didn't capture it.

### CODE REVIEW

**Comment by GrahamSH-LLK** [comment]
This has been denied by @WorldLanguages AFAIK

**Comment by ErrorGamer2000** [comment]
Can you give me a link to it? The first time I tried it I was really new to js, and just threw a
```css
* {
  font-family: xxx !important
}
```
in the code. This is much better.

**Comment by mxmou** [comment]
See https://github.com/ScratchAddons/ScratchAddons/issues/1646#issuecomment-796200979

**Comment by ErrorGamer2000** [comment]
Okay, yes other extensions can do this. This addon makes it customizable for the major parts of scratch, WITHOUT INTERFERING WITH THE TEXT IN THE PAINT EDITOR, which a normal extension that customizes fonts does. It also changes the letter spacing to fit the blocks and prevent overflow. FROM EXPERIENCE, I know that other extensions change the paint editor's font in vector mode because I used to use [this](https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh) to customize fonts.

**Review by Secret-chest — APPROVED** [review]
Good

File: addons/custom-fonts/userscript.js
Comment (WorldLanguages) [inline addons/custom-fonts/userscript.js:2]: Where's webfont?

**Comment by WorldLanguages** [comment]
> Okay, yes other extensions can do this. This addon makes it customizable for the major parts of scratch, WITHOUT INTERFERING WITH THE TEXT IN THE PAINT EDITOR, which a normal extension that customizes fonts does. It also changes the letter spacing to fit the blocks and prevent overflow. FROM EXPERIENCE, I know that other extensions change the paint editor's font in vector mode because I used to use [this](https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh) to customize fonts.

That sounds OK then. As long as the addon has reasonable Scratch specific code it could be added

File: addons/custom-fonts/userscript.js
Comment (WorldLanguages) [reply addons/custom-fonts/userscript.js:2]: You forgot to include this file in the PR, so it won't work

Comment (ErrorGamer2000) [reply addons/custom-fonts/userscript.js:2]: Oh, just a sec.

**Comment by ErrorGamer2000** [comment]
> Set the default to Roboto as Helvetica is not a Google font

I know it isn't but it is the default used by scratch so it will always be loaded. And my script does not try to re-load fonts if they are already available.

File: addons/custom-fonts/userscript.js
Comment (ErrorGamer2000) [reply addons/custom-fonts/userscript.js:2]: I don't have it with me today, I will upload it monday afternoon.

**Comment by Mr-Anonymous75** [comment]
> > Okay, yes other extensions can do this. This addon makes it customizable for the major parts of scratch, WITHOUT INTERFERING WITH THE TEXT IN THE PAINT EDITOR, which a normal extension that customizes fonts does. It also changes the letter spacing to fit the blocks and prevent overflow. FROM EXPERIENCE, I know that other extensions change the paint editor's font in vector mode because I used to use [this](https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh) to customize fonts.
> 
> That sounds OK then. As long as the addon has reasonable Scratch specific code it could be added

If it has been accepted, then why is it nowhere to be found? I really want to see lists and variables in a monospace font...

**Comment by GrahamSH-LLK** [comment]
> > > Okay, yes other extensions can do this. This addon makes it customizable for the major parts of scratch, WITHOUT INTERFERING WITH THE TEXT IN THE PAINT EDITOR, which a normal extension that customizes fonts does. It also changes the letter spacing to fit the blocks and prevent overflow. FROM EXPERIENCE, I know that other extensions change the paint editor's font in vector mode because I used to use [this](https://chrome.google.com/webstore/detail/dark-reader/eimadpbcbfnmbkopoojfekhnkhdbieeh) to customize fonts.
> > 
> > 
> > That sounds OK then. As long as the addon has reasonable Scratch specific code it could be added
> 
> If it has been accepted, then why is it nowhere to be found? I really want to see lists and variables in a monospace font...

Please try to understand how GitHub, and ScratchAddons work. If someone creates a Pull Request, like this, it has to be reviewed and fixed. Once it has been reviewed by the Scratch Addons team, we can merge it. That puts it on our master branch, which won't be released until Monday, June 14th. THEN, you have to wait for Chrome to approve it, and for your device to auto-download it.

File: addons/custom-fonts/userscript.js
Comment (WorldLanguages) [inline addons/custom-fonts/userscript.js:133]: This will not work with `"runAtComplete": false`, body will probably be undefined

**Comment by WorldLanguages** [comment]
Things we need to fix before merging:
- Not using `document.body.appendChild` so that the addon works (I had to change that for it to work). Not sure if `document.head` could be used instead, since the element is appended to then check its `clientWidth`.
- Fallback to `Helvetica` if the provided font name is invalid.
- Clarify font names are case sensitive.
- `trim()` the font name, in case users add an extra space at the end, they wouldn't understand why it doesn't work.
- You should use `innerText` or `textContent` instead of `innerHTML`, unless really needed

**Comment by TheColaber** [comment]
There should also be some example fonts since I've only memorized the sora font's name

**Comment by Mr-Anonymous75** [comment]
> Please try to understand how GitHub, and ScratchAddons work. If someone creates a Pull Request, like this, it has to be reviewed and fixed. Once it has been reviewed by the Scratch Addons team, we can merge it. That puts it on our master branch, which won't be released until Monday, June 14th. THEN, you have to wait for Chrome to approve it, and for your device to auto-download it.

Oh, okay. I didn't realize that a request has to be merged too after being reviewed.

**Review by GrahamSH-LLK — CHANGES_REQUESTED** [review]

File: addons/custom-fonts/userscript.js
Comment (ErrorGamer2000) [reply addons/custom-fonts/userscript.js:4]: Oof, sorry. I will fix it when I get the chance, I am trying to learn Visual Studio Code right now, and it is a very steep learning curve.

**Comment by WorldLanguages** [comment]
I plan on adding a list of suggested fonts. Any suggestions @TheColaber @ErrorGamer2000 ?

**Comment by TheColaber** [comment]
sora.

**Comment by Sly-Little-Fox** [comment]
> I plan on adding a list of suggested fonts. Any suggestions @TheColaber @ErrorGamer2000 ?

It was not for me, but I would add Nunito.

**Comment by Secret-chest** [comment]
> I plan on adding a list of suggested fonts. Any suggestions @TheColaber @ErrorGamer2000 ?

@WorldLanguages Lato, Roboto, Open Sans, Motserrat, Ubuntu.

**Comment by WorldLanguages** [comment]
> I plan on adding a list of suggested fonts.

Update: \<datalist> sucks

**Comment by TheColaber** [comment]
Custom datalist library? :eyes:

**Comment by GrahamSH-LLK** [comment]
> Custom datalist library? 👀

YES! We should code one. I feel like I have a million projects that can use it.

**Comment by WorldLanguages** [comment]
Probably not going to make it for v1.16.0 since we want to figure out a way to suggest font names to the user (but still let it pick any), and \<datalist> doesn't really work for our usecase because it cannot be programmed to show all suggestions when the input is filled.

**Comment by ErrorGamer2000** [comment]
> I plan on adding a list of suggested fonts. Any suggestions @TheColaber @ErrorGamer2000 ?

Sorry it took me so long.

- Roboto
- Open Sans
- PT Sans
- Ubuntu
- Nunito
- Noto Sans

**Comment by Mr-Anonymous75** [comment]
> I plan on adding a list of suggested fonts. Any suggestions @TheColaber @ErrorGamer2000 ?
What about Raleway? Or a monospace font?

**Comment by CubeyTheCube** [comment]
> I plan on adding a list of suggested fonts. Any suggestions?

Helvetica (Neue)
Comic Sans
Roboto Mono

**Comment by WorldLanguages** [comment]
Keeping as a draft until we have a way to list suggested setting values. HTML \<datalist> almost works, but as it is typical, HTML is not good enough for our usecase.

**Comment by TheColaber** [comment]
`<datalist>` isn't *that* bad... I think it would be fine in the settings page, but sure, find a custom data list that you like better.

**Comment by WorldLanguages** [comment]
Maybe we could make use of recently merged #3156 plus presets
But we already have 6 new addons this release so this shouldn't be merged for v1.18

File: addons/custom-fonts/userscript.js
Comment (WorldLanguages) [reply addons/custom-fonts/userscript.js:133]: @ErrorGamer2000 This was never fixed...

**Comment by WorldLanguages** [comment]
> `<datalist>` isn't _that_ bad... I think it would be fine in the settings page, but sure, find a custom data list that you like better.

It is, there's no way to make \<datalist> show all available options when one is already chosen. You need to clear the whole input so that all of them can be seen.

File: addons/custom-fonts/userscript.js
Comment (RedGuy12) [reply addons/custom-fonts/userscript.js:133]: do you even need to append it?

**Comment by RedGuy12** [comment]
Resolves #1646

**Comment by TheColaber** [comment]
@ErrorGamer2000 @WorldLanguages I created a custom data list. Feel free to change the values listed there.

**Comment by ErrorGamer2000** [comment]
I will go back over this addon this afternoon, and see if I can fix the spacing issues. I may have to do some path editing in Blockly which I am reluctant to try because of the block shape and catblocks addons.

**Comment by ErrorGamer2000** [comment]
Hopefully will get it done this afternoon.

**Comment by ErrorGamer2000** [comment]
@TheColaber Will you please look at the changes that i have made to the data lists?

**Comment by TheColaber** [comment]
Looks good, but I'll test it out soon.

**Comment by ErrorGamer2000** [comment]
Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.

**Comment by hello-smile6** [comment]
> Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.

How does that API work?

**Comment by ErrorGamer2000** [comment]
> > Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.
> 
> How does that API work?

https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration

**Comment by hello-smile6** [comment]
> > > Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.
> > 
> > 
> > How does that API work?
> 
> https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration

Is `.gitignore` bloat?
https://github.com/JaydenDev/searchocean/pull/5

**Comment by ErrorGamer2000** [comment]
> > > > Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.
> > > 
> > > 
> > > How does that API work?
> > 
> > https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration
> 
> Is `.gitignore` bloat?
> https://github.com/JaydenDev/searchocean/pull/5

Not sure what you are asking... A `.gitignore` file should only be used to preven committing private, unwanted, or large amounts of files. It is commonly seen with node because there are a lot of files in the node_modules folder.

**Comment by hello-smile6** [comment]
> > > > > Okay, thanks. I am looking into using the `CSSStyleDeclaration` api next instead of setting the innerText of the style, but the styles seem to be interfering. Any help would be appreciated.
> > > > 
> > > > 
> > > > How does that API work?
> > > 
> > > 
> > > https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration
> > 
> > 
> > Is `.gitignore` bloat?
> > [JaydenDev/searchocean#5](https://github.com/JaydenDev/searchocean/pull/5)
> 
> Not sure what you are asking... A `.gitignore` file should only be used to preven committing private, unwanted, or large amounts of files. It is commonly seen with node because there are a lot of files in the node_modules folder.

Okay. Now, could you get @JaydenDev to realize that?

**Comment by ErrorGamer2000** [comment]
Brief me on the repo's purpose please.

**Comment by hello-smile6** [comment]
> Brief me on the repo's purpose please.

A website that lets you quickly do searches with different search engines. I created a NodeJS server for it, but they don't like it.

**Comment by ErrorGamer2000** [comment]
> > Brief me on the repo's purpose please.
> 
> A website that lets you quickly do searches with different search engines. I created a NodeJS server for it, but they don't like it.

That repo looks like a frontend repo. The server should be in a different one.

**Comment by hello-smile6** [comment]
> > > Brief me on the repo's purpose please.
> > 
> > 
> > A website that lets you quickly do searches with different search engines. I created a NodeJS server for it, but they don't like it.
> 
> That repo looks like a frontend repo. The server should be in a different one as it is backend.

I was going to add a server at one point, but they didn't like that. Now I might accidentally commit server-side code.

**Comment by ErrorGamer2000** [comment]
Then delete it, or delete the entire repo on your device then load it again from github.

**Comment by hello-smile6** [comment]
> Then delete it, or delete the entire repo on your device then load it again from github.

That's a bit too drastic.

**Comment by ErrorGamer2000** [comment]
That is really the only way to get rid of all of it...

**Comment by Secret-chest** [comment]
You can add more fonts: Caveat, Handlee, Antonio, Lato, Roboto Mono, Times New Roman

**Comment by ErrorGamer2000** [comment]
> You can add more fonts: Caveat, Handlee, Antonio, Lato, Roboto Mono, Times New Roman

I was planning to add more, but had not had a chance yet.

**Comment by TheColaber** [comment]
I don't need credits, all I did was change the manifest to support data lists, you actually coded the addon.

**Comment by ErrorGamer2000** [comment]
> I don't need credits, all I did was change the manifest to support data lists, you actually coded the addon.

Still, you helped with it.

File: addons/custom-fonts/addon.json
Comment (TheColaber) [inline addons/custom-fonts/addon.json:39]: Sora probably shouldn't be first, that was my fault.

**Comment by TheColaber** [comment]
> > I don't need credits, all I did was change the manifest to support data lists, you actually coded the addon.
> 
> Still, you helped with it.

With the settings page scripts, not the addon. The credit is completely yours for the addon.

**Comment by ErrorGamer2000** [comment]
Fine. I will change it tomorrow

File: addons/custom-fonts/addon.json
Comment (Secret-chest) [reply addons/custom-fonts/addon.json:39]: Yes. It's not a very legible font. Sora should be last.

Comment (ErrorGamer2000) [reply addons/custom-fonts/addon.json]: Fixed it. I did not put it last, but right along with the other fonts that are somewhat unreadable.

**Comment by TheColaber** [comment]
You have one unused variable. https://lgtm.com/projects/g/ScratchAddons/ScratchAddons/rev/pr-b1f552fda7bf7e5e7248c9d332187387e7a34898

File: addons/custom-fonts/userscript.js
Comment (RedGuy12) [inline addons/custom-fonts/userscript.js:28]: textContent

File: addons/custom-fonts/userscript.js
Comment (RedGuy12) [inline addons/custom-fonts/userscript.js:40]: textContent

File: addons/custom-fonts/userscript.js
Comment (RedGuy12) [inline addons/custom-fonts/userscript.js:116]: textContent

**Comment by ErrorGamer2000** [comment]
> You have one unused variable. https://lgtm.com/projects/g/ScratchAddons/ScratchAddons/rev/pr-b1f552fda7bf7e5e7248c9d332187387e7a34898

Ah, yes. I forgot that one.

**Comment by ErrorGamer2000** [comment]
Addon manifest validation fails because of outdated schema. Fixed in ScratchAddons/manifest-schema#53

**Review by TheColaber — DISMISSED** [review]

**Comment by TheColaber** [comment]
I wonder what other addons could use data-list. If there are any then someone can pr that.

**Comment by ErrorGamer2000** [comment]
> I wonder what other addons could use data-list. If there are any then someone can pr that.

Should I add that to this pr?

**Comment by hello-smile6** [comment]
> I wonder what other addons could use data-list. If there are any then someone can pr that.

Maybe <a href="https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratch-notifier"><code>scratch-notifier</code></a>, for the notification sound menu?

**Comment by TheColaber** [comment]
The user can really create custom sounds. And mv3 will prevent sounds played in workers, so the only way would to be to create a new tab and play a sound there. And that's not something we should do either.

**Comment by hello-smile6** [comment]
> The user can really create/load custom sounds. And mv3 will prevent sounds played in workers, so the only way would to be to create a new tab and play a sound there. And that's not something we should do either.

Why not just give the worker script a reference to a random tab's audio-related functions?

**Comment by RedGuy12** [comment]
> Why not just give the worker script a reference to a random tab's audio-related functions?

No guarantee that there'll be a Scratch tab open.

Please stop talking about off-topic things here.

**Comment by hello-smile6** [comment]
> > Why not just give the worker script a reference to a random tab's audio-related functions?
> 
> No guarantee that there'll be a Scratch tab open.
> 
> Please stop talking about off-topic things here.

Sorry.

**Comment by hello-smile6** [comment]
> > Why not just give the worker script a reference to a random tab's audio-related functions?
> 
> No guarantee that there'll be a Scratch tab open.
> 
> Please stop talking about off-topic things here.

Good point. I hid all of my comments that were off-topic.

**Comment by ErrorGamer2000** [comment]
@WorldLanguages could you please review this?

**Comment by ErrorGamer2000** [comment]
@WorldLanguages

**Comment by ErrorGamer2000** [comment]
@WorldLanguages please

**Comment by Endermaster3690** [comment]
> wl please
> wl

probably not gonna if you're spamming him

**Comment by ErrorGamer2000** [comment]
I know that. I wasnt actually expecting it to happen.

**Comment by apple502j** [comment]
We may want to make normal selections use datalist - see #3642
Pending until we make a decision

**Comment by CST1229** [comment]
> datalist

Are those dropdowns? (that aren't the current everything-shown-at-once select options)

**Comment by Samq64** [comment]
> We may want to make normal selections use datalist - see #3642 Pending until we make a decision

I don't think datalist would work well everywhere because users can enter anything they want and it would most likely mess up options that like
image

**Comment by RedGuy12** [comment]
> We may want to make normal selections use datalist - see #3642

datalist is for autocompleting text inputs. it doesn't require that the value is one of the dropdown options, where #3642 is suggesting to change the buttons to `input[type="select"]`s

**Comment by TheColaber** [comment]
> We may want to make normal selections use datalist - see #3642 Pending until we make a decision

Well yes, I'm on board, what is pending? A separate pr can change some normal radio inputs into datalists, as I said here: https://github.com/ScratchAddons/ScratchAddons/pull/2538#issuecomment-1003150981.

**Comment by ErrorGamer2000** [comment]
> > We may want to make normal selections use datalist - see #3642
> 
> datalist is for autocompleting text inputs. it doesn't require that the value is one of the dropdown options, where #3642 is suggesting to change the buttons to `input[type="select"]`s

I agree. Datalist and select options are very different. After looking through the addons, there are definitely some that could use datalist, such as editor-theme-3, and others that should not, like hide-flyout. It really depends on how many options it has

**Comment by apple502j** [comment]
Okay - then it seems good! @ErrorGamer2000 can you fix conflicts (in this PR and schema?)

**Comment by TheColaber** [comment]
@apple502j schema should be from the new setting type.

File: addons/custom-fonts/addon.json
Comment (mxmou) [inline addons/custom-fonts/addon.json:107]: This name is confusing: especially because another setting is called "Footer font", I thought this would apply to the editor menu bar or something else at the top of a page. Box headers are an outdated design element that is only found on old pages (2015 and earlier), so I don't think they need a separate setting. A "heading font" that would apply to all h# elements would be much more useful.

**Comment by ErrorGamer2000** [comment]
Will fix later today.

**Comment by Secret-chest** [comment]
WAIT - do you actually need a userscript?

File: addons/custom-fonts/addon.json
Comment (Secret-chest) [reply addons/custom-fonts/addon.json:121]: > Box headers are an outdated design element that is only found on old pages (2015 and earlier)

Actually, no. They are on the front page, and they got a new class name.

**Comment by Secret-chest** [comment]
> > datalist
> 
> Are those dropdowns? (that aren't the current everything-shown-at-once select options)

No, they are text inputs with presets you can choose (but you can still type what you want)

File: addons/custom-fonts/addon.json
Comment (mxmou) [reply addons/custom-fonts/addon.json:121]: The current front page was designed in 2015.

**Comment by ErrorGamer2000** [comment]
Yes, I need a userscript to calculate sizing for the block editor.

File: addons/custom-fonts/addon.json
Comment (ErrorGamer2000) [reply addons/custom-fonts/addon.json:121]: Yeah, they are on the front page still, but I set it to include all `h#` elements too.

**Comment by WorldLanguages** [comment]
Sorry for the hiatus - so it looks like now the fonts are included inside the addon, right? We'll need to figure out the licensing.
I like the hybrid approach of suggesting fonts as well as letting users type any fonts already on their system (e.g. `Arial Black`)

**Comment by ErrorGamer2000** [comment]
It should work already, but we may need to  remove validation vode from the datalist component.

**Comment by WorldLanguages** [comment]
What does the datalist component actually validate?

**Comment by JaydenDev** [comment]
What are you using for formatting? I sure hope it isn't prettier.

**Comment by CST1229** [comment]
> What are you using for formatting? I sure hope it isn't prettier.

It is.

**Comment by ErrorGamer2000** [comment]
If someone wants to get my attention on here, just @mention me, it pushes a notification on my phone.

**Comment by TheColaber** [comment]
@ErrorGamer2000 The fact that there are 184 items on the data-list means something is wrong. This list was supposed to be short, recommend fonts that actually work and look good in scratch. The new changes also change something we were trying to fix with the native data list element, which was that when you input letters into the input, you can't see the full list anymore. Let's see what @WorldLanguages thinks, but I can tell you, that a list that's 100x bigger than the addon is definitely off.

**Comment by Secret-chest** [comment]
> We'll need to figure out the licensing.

All Google Fonts are under the OFL, which is not compatible with the GPL v3

File: addons/custom-fonts/addon.json
Comment (Secret-chest) [inline addons/custom-fonts/addon.json:15]: Change "letters" to characters

File: addons/custom-fonts/addon.json
Comment (Secret-chest) [inline addons/custom-fonts/addon.json:66]: Care for Linux users :) Don't add Windows-only web-safe fonts (DejaVu Sans is safe as it is included with Windows and Linux)

**Review by Secret-chest — CHANGES_REQUESTED** [review]

File: addons/custom-fonts/addon.json
Comment (mxmou) [reply addons/custom-fonts/addon.json:15]: I don't think "block characters" would be very easy to understand. "Block text" might be better".

**Comment by apple502j** [comment]
@ErrorGamer2000 This doesn't work at all on Firefox:

```
Uncaught (in promise) TypeError: chrome.fontSettings is undefined
    updateItems moz-extension://d2420284-e09f-4bba-b7c0-e5567c11f8d7/webpages/settings/components/addon-setting.js:120
    bind VueJS
    created moz-extension://d2420284-e09f-4bba-b7c0-e5567c11f8d7/webpages/settings/components/addon-setting.js:170
```

**Comment by ErrorGamer2000** [comment]
I will have to wrap it in a `try ... finally` block because it does not look like firefox supports that. Maybe I should hard-code a list of fonts installed by default on most sustems.

**Comment by mxmou** [comment]
> I will have to wrap it in a `try ... finally` block because it does not look like firefox supports that. Maybe I should hard-code a list of fonts installed by default on most sustems.

There is already a hardcoded list for the initial menu - does autofill need to use a different one?

**Comment by ErrorGamer2000** [comment]
I dont quite understand what you mean @mxmou.

**Comment by ErrorGamer2000** [comment]
@apple502j  can you look over this again? It should work in firefox now, but the list just won't fill automatically.

**Comment by Chiroyce1** [comment]
> It should work in firefox now

It doesn't. This error is still there
```
Uncaught (in promise) TypeError: chrome.fontSettings is undefined
    updateItems moz-extension://caf53219-e41b-d74d-b4d7-f5d17b685029/webpages/settings/components/addon-setting.js:159
    bind VueJS
    created moz-extension://caf53219-e41b-d74d-b4d7-f5d17b685029/webpages/settings/components/addon-setting.js:217
    VueJS 6
[addon-setting.js:159:15](moz-extension://caf53219-e41b-d74d-b4d7-f5d17b685029/webpages/settings/components/addon-setting.js)
```

image

And there's also a problem with the manifest file

```
Reading manifest: Warning processing permissions: Error processing permissions.9: Value "fontSettings" must either: must either [must either [must either [must either [be one of ["idle"], be one of ["menus.overrideContext"], be one of ["search"], be one of ["activeTab"], be one of ["cookies"], or be one of ["webRequest", "webRequestBlocking"]], be one of ["clipboardRead", "clipboardWrite", "geolocation", "notifications"], be one of ["bookmarks"], be one of ["browsingData"], be one of ["devtools"], be one of ["find"], be one of ["history"], be one of ["pkcs11"], be one of ["sessions"], be one of ["tabs", "tabHide"], be one of ["topSites"], be one of ["browserSettings"], be one of ["downloads", "downloads.open"], be one of ["management"], be one of ["privacy"], be one of ["proxy"], be one of ["nativeMessaging"], or be one of ["webNavigation"]], be one of ["alarms", "mozillaAddons", "storage", "unlimitedStorage"], be one of ["captivePortal"], be one of ["identity"], be one of ["menus", "contextMenus"], be one of ["normandyAddonStudy"], be one of ["geckoProfiler"], be one of ["urlbar"], be one of ["contextualIdentities"], be one of ["dns"], be one of ["activityLog"], be one of ["networkStatus"], be one of ["telemetry"], or be one of ["theme"]], must either [must either [be one of ["idle"], be one of ["menus.overrideContext"], be one of ["search"], be one of ["activeTab"], be one of ["cookies"], or be one of ["webRequest", "webRequestBlocking"]], be one of ["clipboardRead", "clipboardWrite", "geolocation", "notifications"], be one of ["bookmarks"], be one of ["browsingData"], be one of ["devtools"], be one of ["find"], be one of ["history"], be one of ["pkcs11"], be one of ["sessions"], be one of ["tabs", "tabHide"], be one of ["topSites"], be one of ["browserSettings"], be one of ["downloads", "downloads.open"], be one of ["management"], be one of ["privacy"], be one of ["proxy"], be one of ["nativeMessaging"], or be one of ["webNavigation"]], or match the pattern /^experiments(\.\w+)+$/], or must either [be one of ["<all_urls>"], must either [match the pattern /^(https?|wss?|file|ftp|\*):\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$/, or match the pattern /^file:\/\/\/.*$/], or match the pattern /^resource:\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$|^about:/]
```

**Comment by ErrorGamer2000** [comment]
Looks like it is completely iseless and broken on firefox then. Maybe I should just remove the autofill.

**Comment by RedGuy12** [comment]
don't try/catch it, do `if(typeof chrome.fontSettings === "undefined") { /* firefox */ } else { /* chrome */ } `

File: addons/custom-fonts/addon.json
Comment (RedGuy12) [inline addons/custom-fonts/addon.json:3]: `Customize the font used on different parts of the Scratch website.`

File: addons/custom-fonts/addon.json
Comment (RedGuy12) [inline addons/custom-fonts/addon.json:4]: community, not editor

File: manifest.json
Comment (RedGuy12) [inline manifest.json:51]: can this not be an optional permission?

Comment (RedGuy12) [reply manifest.json:51]: also, update the workflow to remove it from the firefox manifest

**Review by RedGuy12 — CHANGES_REQUESTED** [review]

**Comment by ErrorGamer2000** [comment]
Done @RedGuy12

**Comment by RedGuy12** [comment]
@ErrorGamer2000 
https://github.com/ScratchAddons/ScratchAddons/blob/61a7af328bbfd3efac6d4e5eddaeaed973ad2c14/background/imports/util.js#L1-L2
https://github.com/ScratchAddons/ScratchAddons/blob/61a7af328bbfd3efac6d4e5eddaeaed973ad2c14/webpages/settings/index.js#L57-L58
Both of these should be followed by `else browserLevelPermissions.push("fontSettings"); `

Also you need to add `"permissions": ["fontSettings"]` to the addon manifest.

That should make it work - remember to test.

File: addons/addons.json
Comment (RedGuy12) [inline addons/addons.json:106]: add this back

**Comment by RedGuy12** [comment]
Please don't ship another copy of Sora.

File: addons/addons.json
Comment (ErrorGamer2000) [reply addons/addons.json:106]: Oh, will do in a little while.

**Comment by TheColaber** [comment]
Closing til we have a complete implementation. I don't think we need to use the chrome.apis. I think we should just use a short list of fonts that are generally used. If someone wants to recreate this, we could start with a pr that just implements the data-list setting and another with this addon.

**Comment by mxmou** [comment]
> Closing til we have a complete implementation. I don't think we need to use the chrome.apis. I think we should just use a short list of fonts that are generally used. If someone wants to recreate this, we could start with a pr that just implements the data-list setting and another with this addon.

If someone reimplements this, there are a few things that could be improved:
* Separate addons for editor and website to make them easier to find on the setting page.
* One setting for the website, one for the editor and one for blocks is probably enough.
* The font stack used by Scratch 2 should be one of the options.
* Redraw blocks instead of changing the font size.

**Pull request closed without merging**
